### ISSUE TITLE
Question Regarding Training Process

### ISSUE THREAD

**Issue by TRvlvr** [tsurumeso/vocal-remover#18]
Hello again!

I have a question about the training process. I'm currently training a model with 170 pairs and I'm currently on my 29th epoch after about 16 hours - 

_# epoch 29
  * inner epoch 0
    * training loss = 1.303589, validation loss = 1.348374
    * best validation loss
  * inner epoch 1
    * training loss = 1.296407, validation loss = 1.359492
  * inner epoch 2
    * training loss = 1.291624, validation loss = 1.358964
  * inner epoch 3
    * training loss = 1.288329, validation loss = 1.359871_

My understanding is the training loss & validation loss should be close to each other, with the validation loss always needing to be slightly higher than the training loss.

Here are my questions -

1. What is the ideal training loss & validation loss I should be looking out for?
2. I noticed new the "model_iter.pth" files spawn after most epochs complete training. Do I need to do anything to the last the "model_iter.pth" file when training runs to the last set epoch (in my case 100 epochs)?
3. I noticed all "model_iter.pth" files are the same size, 37,830KB's, but your newest model is 53,130KB's. Does a larger model size mean better training, or will it increase with a larger dataset?
4. How do I increase the size of the model?
5. Sometimes I don't get a "model_iter.pth" model when an epoch finishes, is there a way I can set it so that I do after every epoch?

Thank you so much in advance!

**Comment by TRvlvr** [comment]
After much research and testing, here is what I gather regarding my questions -

1. The goal is to get the training loss & validation loss as low as possible. I noticed the lower the validation loss, the better results I get on on conversions.
2. My assumption is that I should wait until I've gotten to the lowest possible training & validation loss number. When I reach a point where the training & validation loss numbers no longer decrease after more than 3 or 4 epochs, it's likely it won't get any better and training has likely completed.
3 & 4. It seems the settings regarding the model size are made in the "net.py" located in the lib directory. The size of the dataset doesn't appear to have an impact.
5. The current settings are most efficient. A new "model-iter.pth" file spawns only if the best possible validation loss has been reported. That way you're not flooded with poor performing models.

**Comment by tsurumeso** [comment]
Sorry for the late reply.

> 1. Can you confirm if my answers above are correct?

Your answer is correct.

> 2. I'm currently training on a set of 477 aligned pairs, 377 training & 100 for validation. How low should the training & validation loss numbers get before I can consider the model effective and complete?

There are no clear threshold, as the losses vary depending on the songs in dataset (male or female vocal, anime songs or pop songs, quality of training data, and so on).
Maybe the following is the answer for this question.

> When I reach a point where the training & validation loss numbers no longer decrease over the course of 3 or 4 epochs, it's likely it won't get any better and the model has reached it's limit.

**Comment by aufr33** [comment]
> _# epoch 29
> 
>     * inner epoch 0
>       * training loss = 1.303589, validation loss = 1.348374

Wow. You probably have flaws in the dataset. At the epoch 29, my losses were about 0.6-0.7.
Make sure the instrumental and mix are perfectly in sync. To check this, I used a simple script that runs **sox** in the following way:

sox --combine mix -v 1 "mixtures\01_mix.wav" -v -1 "instrumentals\01_inst.wav" "01_vocals.wav"

Then open each *_vocals.wav file in Adobe Audition, go to **Show Spectral Frequency Display** (Shift + D) and listen. Vocals must be clean, no instruments should be heard.

**Comment by TRvlvr** [comment]
> > _# epoch 29
> > ```
> > * inner epoch 0
> >   * training loss = 1.303589, validation loss = 1.348374
> > ```
> 
> Wow. You probably have flaws in the dataset. At the epoch 29, my losses were about 0.6-0.7.
> Make sure the instrumental and mix are perfectly in sync. To check this, I used a simple script that runs **sox** in the following way:
> 
> sox --combine mix -v 1 "mixtures\01_mix.wav" -v -1 "instrumentals\01_inst.wav" "01_vocals.wav"
> 
> Then open each *_vocals.wav file in Adobe Audition, go to **Show Spectral Frequency Display** (Shift + D) and listen. Vocals must be clean, no instruments should be heard.

I'm glad you brought this up! 

I already did this with a similar script that allowed me to perfectly align mixtures and official instrumentals. What my problem was I didn't consider the fact that the wave-forms should also align. The volumes should be identical. If the subtracted track has trace amounts of audible instrumentation, but the vocals are clearly higher, it's likely the wave forms didn't align. I ended up using too many pairs with misaligned wave-forms. Using multi-tracks to create pairs is the most concrete way to build a perfectly aligned dataset. However, carefully aligning wave-forms and spectrograms is a pretty good way too because you can easily expand your dataset.

I trained with 550 pairs and got GREAT results.

**Comment by nkcdy** [comment]
I use dataset DSD100 for the training. the initial loss is much smaller than yours:

# epoch 5
  * inner epoch 0
    * training loss = 0.000357, validation loss = 0.000336
  * inner epoch 1
    * training loss = 0.000350, validation loss = 0.000339
  * inner epoch 2
    * training loss = 0.000347, validation loss = 0.000307
    * best validation loss
  * inner epoch 3
    * training loss = 0.000343, validation loss = 0.000308

I'm wondering if it is normal.

**Comment by aufr33** [comment]
@nkcdy This is because in the old version (2.2) the values were multiplied by 1000.

**Comment by nkcdy** [comment]
> @nkcdy This is because in the old version (2.2) the values were multiplied by 1000.

the loss value is still very small even mutiplied by 1000.
